### PR TITLE
[WIP] Preliminary support for Text Editor option pages

### PR DESCRIPTION
**Pull request by majocha**
Uses `SVsSettingsPersistenceManager` to persist settings.

Styling follows default Roslyn look.

UI can be created in code with some helpers or provided as a WPF control.

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
this is great :)

**Comment by cloudRoutine** [comment]
@majocha just use blend 😉

File: vsintegration/src/FSharp.Editor/Options/IntelliSenseOptions.fs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+module IntelliSenseOptionPageStrings =
+    let Option_CompletionLists = SR.GetString "Completion_Lists"
+    let Option_Show_completion_list_after_a_character_is_typed = SR.GetString "Show_completion_list_after_a_character_is_typed"
+    let Option_Show_completion_list_after_a_character_is_deleted = SR.GetString "Show_completion_list_after_a_character_is_deleted"
+
+type IntelliSenseOptionPageControl(serviceProvider) as this =
+    inherit AbstractOptionPageControl(serviceProvider)
+
+    let ( *** ) (control : 'T when 'T :> IAddChild) (children: UIElement list) =
```
Comment (cloudRoutine) [inline vsintegration/src/FSharp.Editor/Options/IntelliSenseOptions.fs]: it'd be useful to have move all the UI related helper functions to a module so the other options pages can eventually use them too

File: vsintegration/src/FSharp.Editor/Common/LanguageService.fs [hunk 2]
```diff
@@ -195,6 +195,7 @@ type internal FSharpCheckerWorkspaceServiceFactory
 
 type
     [<Guid(FSharpCommonConstants.packageGuidString)>]
+    [<ProvideLanguageEditorOptionPage(typeof<Options.IntelliSenseOptionPage>, "F#", null, "IntelliSense", "6008")>]
```
Comment (cloudRoutine) [inline vsintegration/src/FSharp.Editor/Common/LanguageService.fs]: How do you know/Where did you find the numbers(keys?) we're supposed to use for these? I'd like to get started on making a few of them too

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Common/LanguageService.fs]: @cloudRoutine search the solution for such numbers. They sit somewhere in a res file, I've added one when implementing the initial intellisense prop page.

**Comment by majocha** [comment]
Seeing that it won't build with `Microsoft.VisualStudio.LanguageServices.Implementation` I'll rewrite it with `UIElementDialogPage`. It looked like a good idea to use all the Roslyn abstractions but practice shows they are not yet intended for general use. Some things are internal like `IOptionService` and lots of them are best suited for C#/VB and overly complicated for our needs.

About the XAML fiasco,
The approach I took was setting XAML compile action to "Resource" and loading it at run time with FsXaml `InjectXaml.from` helper. It did work in simple case but I was hoping to mostly copy/paste [IntelliSenseOptionPageControl.xaml](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.CSharp/Options/IntelliSenseOptionPageControl.xaml) The problem is localization. I know next to nothing about this, so I tried to just copy what Roslyn does, i.e. point to a module/static class with `Content={x:Static... }` but couldn't get it to work.
This, and the fact that still more modifications were needed to load the XAML, and the need to manually track all the `x:Name` elements tipped the scales. I realized I'm creating unmaintainable mess and backed up.

On second thought it's still feasible to use XAML, just not the one from Roslyn :)

File: vsintegration/src/FSharp.Editor/Common/LanguageService.fs (hunk 2 above)
Comment (majocha) [inline vsintegration/src/FSharp.Editor/Common/LanguageService.fs]: @cloudRoutine yes, I used the one Vasily created in FSharp.Editor resources.

**Comment by cloudRoutine** [comment]
Sure that works, I'm fine with writing WPF UI elements straight in code if it comes to that. As long as it isn't winforms it works for me

**Comment by dsyme** [comment]
This is fantastic

**Comment by majocha** [comment]
It will take some time, because we will need some persistence solution for settings.
I see that everybody uses the officially not supported `SVsSettingsPersistenceManager`:
[RTVS](https://github.com/Microsoft/RTVS/blob/f6d7effc6f58067af50b1c5db68b725f2bde1af0/src/Package/Impl/Shell/VsSettingsStorage.cs#L23)

[Roslyn](https://github.com/dotnet/roslyn/blob/f9cfa41ace5035f13063789e049da2ed0ed0370d/src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs#L31)

We could probably port and adapt the RTVS implementation.

**Comment by majocha** [comment]
Integration with CompletionProvider now works.
Some resilience and clean up to do.

**Comment by vasily-kirichenko** [comment]
It's great :)

**Comment by majocha** [comment]
God, this is awful. @cloudRoutine, I need help merging your refactoring.

**Comment by vasily-kirichenko** [comment]
I think it was not right to merge a newer, large, pure refactoring PR having several other PR waiting for merging for a while.

**Comment by majocha** [comment]
This is hopeless. I'll close it for now and make a good PR when it's done.

**Comment by smoothdeveloper** [comment]
Is there a way to tell if merging a PR would occasion other PR to be conflicting?

It sounds like something github should focus on.

**Comment by majocha** [comment]
Something is borked on GitHub because it shows as changes things that definitely are not. Anyway I'll try to "replay" this onto current master. It was just 3 files + some very minor changes. It's almost done. Already works, just some more tests needed.

**Comment by KevinRansom** [comment]
@vasily-kirichenko we always have PR's as WIP or waiting to merge.  If we want to allow refactoring of the code to "improve it" then it's a tax that will have to be paid.  
There is an alternative approach to consider and that is require feature PRs to be minimum necessary diffs and disallow refactoring.
Personally, I'm okay with the tax of refactoring PR's with the result of a better organized code-base.  But I am willing to be guided into an alternate approach if you think that's better.

Kevin

**Comment by cloudRoutine** [comment]
@majocha continue with https://github.com/cloudRoutine/visualfsharp/tree/options-no-xaml

in this instance a `git pull Microsoft master --rebase` would have served you better than merging directly

@smoothdeveloper that's what `git fetch` is for

**Comment by majocha** [comment]
Thanks @cloudRoutine! I need to improve my git skills :)

**Comment by smoothdeveloper** [comment]
@KevinRansom I think we are fine with getting the refactors in, in this case we can use the trick of adding obsolete type aliases, which would have helped / would still help pending PR that aren't rebased yet.

**Comment by cloudRoutine** [comment]
@smoothdeveloper do you have any idea how long that would take? I could fix every outstanding PR and then some faster than going through every function manually adding the attribute, redirecting the implementation to another one, which often won't be feasible if the ordering has changed, requires leaving lots of files that are supposed to be deleted and gets the repo stuck in an interstitial mess until however many open PRs are updated or closed.

just using git properly is a lot simpler.

**Comment by smoothdeveloper** [comment]
Ok, I had the renamed types in mind in my comment, not mistakes about using git wrongly.

**Pull request closed without merging**